antsibull/docsite_docs.py

```diff
@@ -0,0 +1,234 @@
+        return None, errors
+    return Section(yaml_section['title'], toctree), errors
+
+
+def load_docsite_index(index_path: str) -> t.Tuple[t.List[Section], t.List[str]]:
+    '''Load a docsite extra-docs.yml file.
+
+    :arg index_path: Path to extra-docs.yml (does not need to exist).
+    :arg report_not_existing: Whether to report when extra-docs.yml does not exist.
+    :returns: A tuple consisting of a list of sections and a list of error messages.
+    :raises: DocsiteIndexError if extra-docs.yml does not exist
+    '''
+    sections: t.List[Section] = []
+    errors: t.List[str] = []
+
+    if not os.path.isfile(index_path):
+        raise DocsiteIndexError('extra-docs.yml does not exist')
+
+    try:
+        index = load_yaml_file(index_path)
+        if index.get('sections'):
+            for section_index, yaml_section in enumerate(index['sections']):
+                if not isinstance(yaml_section, dict):
+                    errors.append(f'Section #{section_index} must be a mapping')
+                    continue
+                section, section_errors = load_section(yaml_section, section_index)
+                if section is not None:
+                    sections.append(section)
+                errors.extend(section_errors)
+    except Exception as exc:
+        errors.append(str(exc))
+
+    return sections, errors
+
+
+async def load_collection_docsite(collection_name: str,
+                                  collection_metadata: AnsibleCollectionMetadata,
+                                  path_prefix: str = 'docsite/'
+                                  ) -> CollectionDocsiteInfoT:
+    '''Given a collection name and collection metadata, load docsite data.
+
+    :arg collection_name: Dotted collection name.
+    :arg collection_metadata: Collection metadata object.
+    :arg path_prefix: Prefix to add to relative paths, and toctree entries.
+    :returns: A tuple consisting of a list of sections and a list of RST documents as tuples
+              (relative path in docsite, content).
+    '''
+    flog = mlog.fields(func='load_collection_docsite')
+    flog.debug('Enter')
+
+    index_path = os.path.join(collection_metadata.path, 'docs', 'docsite', 'extra-docs.yml')
+    try:
+        sections, _ = load_docsite_index(index_path)
+    except DocsiteIndexError:
+        sections = []
+
+    for section in sections:
+        for i, toctree in enumerate(section.toctree):
+            section.toctree[i] = path_prefix + toctree
+    documents = []
+    for doc in find_docsite_docs(collection_metadata.path):
+        try:
+            # Load content
+            async with aiofiles.open(doc[0], 'r', encoding='utf-8') as f:
+                content = await f.read()
+
+            # Lint content
+            _, errors = lint_required_conditions(content, collection_name)
+
+            # When no errors were found, add to output
+            if not errors:
+                documents.append((path_prefix + doc[1], content))
```

Comment:
I'm a little worried about how much additional memory this might end up using but I'm not sure if I need to worry about that now or if it can be deferred until it actually becomes a problem.  I can see three strategies we could take:

* What you're doing.  We only read the files once.  We validate the data prior to further processing.
* Validate when we use the data (ie, when we copy it over to the docsite location).  This has the benefit of only having the data in memory when it is copied.  The drawback is that we don't know if the pages have valid data until we're ready to write it.
* Discard the data after validating and then read it again when we copy.  Benefits of both the above approaches.  Drawback is that we read the data twice which could be slower if the files aren't cached in memory by the OS.

We could also implement our own cache (something as simple as @functools.lru_cache() on a function which returns the contents of a full path to a file could work for small sets of extra_docs but it will probably have more cache misses than if we code something ourselves) to limit memory usage while also reading less than the full data set each time.

Do you think we should change this strategy now or wait until we start running into problems?

Reply:
> Discard the data after validating and then read it again when we copy.

I think that's the best approach. During a regular docs build, the chance that data changes between the two reads should be 0.

Reply:
Implemented in b927800e6798f18b6692e9d930009486096fdaf9.